File: 卡码/代码随想录/Go/动态规划/05-子序列问题/01-子序列(不连续)/LeetCode_1035.go
// 1035. 不相交的线
package main

import (
	"math"
)

func maxUncrossedLines(nums1 []int, nums2 []int) int {
	dp := make([][]int, len(nums1)+1)
	for i := 0; i < len(nums1)+1; i++ {
		dp[i] = make([]int, len(nums2)+1)
	}
	for i := 1; i < len(nums1)+1; i++ {
		for j := 1; j < len(nums2)+1; j++ {
			if nums1[i-1] == nums2[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				dp[i][j] = int(math.Max(float64(dp[i][j-1]), float64(dp[i-1][j])))
			}
		}
	}
	return dp[len(nums1)][len(nums2)]
}

File: 卡码/代码随想录/1/09-动态规划/Java/dp/src/main/java/com/kurumi/dp/_05_子序列问题/_01_不连续子序列/LeetCode_1035.java
//1035. 不相交的线
package com.kurumi.dp._05_子序列问题._01_不连续子序列;

public class LeetCode_1035 {

}
class Solution {
  public int maxUncrossedLines(int[] nums1, int[] nums2) {
    int[][] dp=new int[nums1.length+1][nums2.length+1];
    for(int i=1;i<nums1.length+1;i++)
      for(int j=1;j<nums2.length+1;j++)
        if(nums1[i-1]==nums2[j-1]) dp[i][j]=dp[i-1][j-1]+1;
        else dp[i][j]=Math.max(dp[i][j-1], dp[i-1][j]);
    return dp[nums1.length][nums2.length];
  }
}
